go/proxyd/daisychain.go

```diff
@@ -0,0 +1,528 @@
+	},
+	"eth_getTransactionCount": []reflect.Type{
+		reflect.TypeOf(common.Address{}),
+		reflect.TypeOf(rpc.BlockNumberOrHash{}),
+		reflect.TypeOf(&RequestOptions{}),
+	},
+}
+
+func NewDaisyChainServer(urls []string) *DaisyChainServer {
+	if len(urls) != 6 {
+		panic("must pass 6 urls")
+	}
+
+	srv := DaisyChainServer{
+		maxBodySize:  100_000, // TODO: from config
+		epoch1RPCURL: urls[0],
+		epoch2RPCURL: urls[1],
+		epoch3RPCURL: urls[2],
+		epoch4RPCURL: urls[3],
+		epoch5RPCURL: urls[4],
+		epoch6RPCURL: urls[5],
+	}
+
+	srv.client = &http.Client{
+		Timeout: 5 * time.Second,
+	}
+
+	return &srv
+}
+
+func StartDaisyChain(config *DaisyChainConfig) (func(), error) {
+	epoch1RPCURL := config.Backends.Epoch1RPCURL
+	epoch2RPCURL := config.Backends.Epoch2RPCURL
+	epoch3RPCURL := config.Backends.Epoch3RPCURL
+	epoch4RPCURL := config.Backends.Epoch4RPCURL
+	epoch5RPCURL := config.Backends.Epoch5RPCURL
+	epoch6RPCURL := config.Backends.Epoch6RPCURL
+
+	urls := []string{
+		epoch1RPCURL,
+		epoch2RPCURL,
+		epoch3RPCURL,
+		epoch4RPCURL,
+		epoch5RPCURL,
+		epoch6RPCURL,
+	}
+
+	defined := false
+	for i, url := range urls {
+		if url != "" {
+			log.Info("epoch rpc url defined", "epoch", i+1)
+			defined = true
+		}
+	}
+	if !defined {
+		panic("must define one epoch url")
+	}
+
+	// parse the config
+	srv := NewDaisyChainServer(urls)
+
+	if config.Metrics.Enabled {
+		addr := fmt.Sprintf("%s:%d", config.Metrics.Host, config.Metrics.Port)
+		log.Info("starting metrics server", "addr", addr)
+		go http.ListenAndServe(addr, promhttp.Handler())
+	}
+
+	// To allow integration tests to cleanly come up, wait
+	// 10ms to give the below goroutines enough time to
+	// encounter an error creating their servers
+	errTimer := time.NewTimer(10 * time.Millisecond)
+
+	if config.Server.RPCPort != 0 {
+		go func() {
+			if err := srv.RPCListenAndServe(config.Server.RPCHost, config.Server.RPCPort); err != nil {
+				if errors.Is(err, http.ErrServerClosed) {
+					log.Info("RPC server shut down")
+					return
+				}
+				log.Crit("error starting RPC server", "err", err)
+			}
+		}()
+	}
+
+	if config.Server.WSPort != 0 {
+		go func() {
+			if err := srv.WSListenAndServe(config.Server.WSHost, config.Server.WSPort); err != nil {
+				if errors.Is(err, http.ErrServerClosed) {
+					log.Info("WS server shut down")
+					return
+				}
+				log.Crit("error starting WS server", "err", err)
+			}
+		}()
+	}
+
+	<-errTimer.C
+	log.Info("started daisychain")
+
+	return func() {
+		log.Info("shutting down daisychain")
+		srv.Shutdown()
+		log.Info("goodbye")
+	}, nil
+}
+
+// TODO: batch support
+func (s *DaisyChainServer) HandleRPC(w http.ResponseWriter, r *http.Request) {
+	ctx := s.populateContext(w, r)
+	if ctx == nil {
+		return
+	}
+
+	log.Info(
+		"received RPC request",
+		"req_id", GetReqID(ctx),
+		"auth", GetAuthCtx(ctx),
+		"user_agent", r.Header.Get("user-agent"),
+	)
+
+	body, err := ioutil.ReadAll(io.LimitReader(r.Body, s.maxBodySize))
```

Comment:
Does `r.Body` needs to `Close`d here?

Reply:
Yes, it does. The linter will catch that one we merge the CircleCI pr.